### DISCUSSION TITLE
android modules and product falvors

### DISCUSSION THREAD

**Discussion by hsynkunduz** [codemagic-ci-cd/codemagic-docs#2780] [Q&A] [answered]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Android (Kotlin)

### Steps to reproduce

tried to search for how to build selecting modules and product flavors on youtube channel, blog, docs, dicord. I could not find anything to show how to do it. Am I missing something somewhere? if there is doc about it how can I do it?

### Expected results

finding sources which shows me how to implement android modules and product flavors on yaml file

### Actual results

there is no information about how to do it

**Comment by dam-ease** [comment]
Hello @hsynkunduz.
You can check out our sample projects here: https://docs.codemagic.io/yaml-quick-start/codemagic-sample-projects/, there is one for building with flavors but with flutter here: https://github.com/codemagic-ci-cd/codemagic-sample-projects/tree/main/flutter/flutter-flavors-demo-project that'd help a whole lot!
If you have specific questions or issues, kindly share here and I'd be glad to help out.

**Comment by hsynkunduz** [comment]
yes I saw those links, I also found blog posts and videos about flutter. But thats flutter flavors, its very different than android ones. Even if it's same still thats not including android modules. I have an android project which have two modules and they have two flavors each. With this I can create two apps and two versions for each of them. I'm trying to get builds I'm able to get from android studio.

**Reply by dam-ease** [reply, marked as answer]
Hello @hsynkunduz
I am currently working on generating a sample project and updating the docs to include building selected modules and product flavors in Android - https://github.com/codemagic-ci-cd/codemagic-docs/issues/2784. I would like to keep you in the loop so you can as well try on your end.
Firstly, creating the flavors in the build.gradle.kts file using the code snippet below
<details><summary>Code Snippet (build.gradle.kts file)</summary>
<p>

```
android {
    ...
    defaultConfig {...}
    buildTypes {
        getByName("debug"){...}
        getByName("release"){...}
    }
    // Specifies one flavor dimension.
    [flavorDimensions](https://developer.android.com/reference/tools/gradle-api/current/com/android/build/api/dsl/ProductFlavor#dimension) += "version"
    productFlavors {
        create("demo") {
            // Assigns this product flavor to the "version" flavor dimension.
            // If you are using only one dimension, this property is optional,
            // and the plugin automatically assigns all the module's flavors to
            // that dimension.
            dimension = "version"
            applicationIdSuffix = ".demo"
            versionNameSuffix = "-demo"
        }
        create("full") {
            dimension = "version"
            applicationIdSuffix = ".full"
            versionNameSuffix = "-full"
        }
    }
}
``` 

</p>
</details> 

with this, we'll have different build variants which can be used to run the app both locally and in codemagic yaml file. To build a specific flavor using the command line, we can use the following Gradle commands:
```
./gradlew demoRelease
```
and to build specific modules in a multi-module Android project, we can specify the module in the Gradle command:
```
./gradlew :moduleName:demoRelease
```

Such that to configure this in Codemagic using the YAML file, we can specify the flavor and module in the build script like this:
```
scripts:
  - name: Build APK with selected flavor
    script: |
      ./gradlew :moduleName:demoRelease
```

**Reply by hsynkunduz** [reply]
Actually I solved this with a little more digging to gradle and how to build android apps with command line. Your solution is very similar to mine. BTW if you are gonna make a doc about this I think you should add build types too. It is very similar so it would not increase your work a lot. I can help if you want. I am now able to build any module, product flavor and build type for android on codemagic.

**Reply by dam-ease** [reply]
Hello @hsynkunduz,
Thanks for your response!
Yes, definitely, I'd be glad to get your input. You can open a PR in the docs repo - https://github.com/codemagic-ci-cd/codemagic-docs or sample project repo - https://github.com/codemagic-ci-cd/codemagic-sample-projects/ in whichever capacity you'd like to help.

**Reply by hsynkunduz** [reply]
I never wrote a documentation like this. But I can add a sample project under android. I think a new android project would be better in here
`https://github.com/codemagic-ci-cd/codemagic-sample-projects/tree/main/android`

Because other android projects have different focus. So I don't want to mix them. What do you think?

**Reply by dam-ease** [reply]
Yes, adding it under android project is fine!

**Reply by dam-ease** [reply]
Hey @hsynkunduz,
Any update for me on this?
Please let me know how it's going on your end.

**Reply by hsynkunduz** [reply]
I was thinking to do this on weekend. But another thing came up. I am under the impression of this has no urgency. So I postponed this to this weekend. If you think this is urgent please let me know.

**Reply by dam-ease** [reply]
Hey @hsynkunduz.
Weekend is very fine, just checking in to know if there were any blockers

**Comment by 1hsyn1** [comment]
Hi @dam-ease I'm hsynkunduz with another github account. I'm ready with sample project in codemagic-sample-projects but since I do not hold permissions I'm not able to push. What should I do.

**Comment by 1hsyn1** [comment]
btw so sorry for the really late response

**Reply by dam-ease** [reply]
Hello @1hsyn1,
No problem at all. Apologies for my late response as well.
You can fork the sample projects repo - https://github.com/codemagic-ci-cd/codemagic-sample-projects, clone the project and add the sample project you have under the android directory - https://github.com/codemagic-ci-cd/codemagic-sample-projects/tree/main/android. After that, kindly push your changes and raise a pull request.
